keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,245 @@
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [Test Framework](#test-framework)
+   * [Test Rollout](#test-rollout)
+   * [Test Portability](#test-portability)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+   * [WIP Future Test Cases](#wip-future-test-cases)
+      * [Pod Teardown](#pod-teardown)
+      * [PV Binding Tests](#pv-binding-tests)
+      * [PV Provisioning Tests](#pv-provisioning-tests)
+      * [PV Deletion Tests](#pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+   * [Phase 3](#phase-3)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### Test Framework
+
+The tests should be developed and run in sig-scalability’s test
+[Cluster Loader framework](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/README.md)
+and infrastructure.
+
+The framework already supports measurements for Pod startup latency and can be
+used to determine a [Pod startup SLO with
+volumes](https://github.com/kubernetes/community/pull/3242).
+
+Additional measurements can be added to the framework to get informational details for
+for volume operations. These are not strictly required but are nice to have:
+
+* Gather existing [volume operation
+  metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/metrics.go)
+  from kube-controller-manager and kubelet.
+  * Caveat: These metrics measure the time of a single operation attempt and not
+    the e2e time across retries. These metrics may not be suitable to measure
+    latency from a pod's perspective.
+* Enhance the scheduler metrics collection to include
+  [volume scheduling
+  metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go)
+* Timing of volume events. However, events should not be considered a reliable
+  measurement because they can be dropped or garbage collected. Also there are
```

Comment:
For stateless pod, we are using events as debugging info - if they are we report them, if not, well - we just have e2e value (which is e2e SLI, which is enough to check if we meet SLO - just harder/impossible to debug in that case).